File: python/toko.py
from produk import Produk
from pembeli import Pembeli
from transaksi import Transaksi

class Toko:
    def __init__(self):
        # List untuk menyimpan semua produk
        self.daftar_produk = []
        # List untuk menyimpan semua pembeli
        self.daftar_pembeli = []
        # List untuk menyimpan semua transaksi
        self.daftar_transaksi = []

    # ===== PRODUK =====
    # Menambahkan objek produk baru ke daftar produk.
    def tambah_produk(self, produk):
        self.daftar_produk.append(produk)
        print("Produk berhasil ditambahkan.\n")

    # Menampilkan semua produk (pakai for-each loop).
    def tampilkan_semua_produk(self):
        for p in self.daftar_produk:
            p.display_info()

    # Mencari produk berdasarkan nama.
    def cari_produk(self, nama):
        found = False
        i = 0
        while i < len(self.daftar_produk) and not found:
            if self.daftar_produk[i].get_nama() == nama:
                self.daftar_produk[i].display_info()
                found = True
            i += 1
        if not found:
            print(f"Produk {nama} tidak ditemukan.\n")

    # Mengubah data produk berdasarkan ID.
    def ubah_produk(self, id_produk, produk_baru):
        found = False
        i = 0
        while i < len(self.daftar_produk) and not found:
            if self.daftar_produk[i].get_id() == id_produk:
                self.daftar_produk[i] = produk_baru
                print("Produk berhasil diubah.\n")
                found = True
            i += 1
        if not found:
            print(f"Produk dengan ID {id_produk} tidak ditemukan.\n")

    # Menghapus produk berdasarkan ID.
    def hapus_produk(self, id_produk):
        found = False
        i = 0
        while i < len(self.daftar_produk) and not found:
            if self.daftar_produk[i].get_id() == id_produk:
                del self.daftar_produk[i]
                print("Produk berhasil dihapus.\n")
                found = True
            i += 1
        if not found:
            print(f"Produk dengan ID {id_produk} tidak ditemukan.\n")

    # ===== PEMBELI =====
    # Menambahkan pembeli baru ke daftar pembeli.
    def tambah_pembeli(self, pembeli):
        self.daftar_pembeli.append(pembeli)
        print("Pembeli berhasil ditambahkan.\n")

    # Menampilkan semua pembeli (pakai for-each loop).
    def tampilkan_semua_pembeli(self):
        for pb in self.daftar_pembeli:
            pb.display_info()

    # Mencari pembeli berdasarkan nama.
    def cari_pembeli(self, nama):
        found = False
        i = 0
        while i < len(self.daftar_pembeli) and not found:
            if self.daftar_pembeli[i].get_nama() == nama:
                self.daftar_pembeli[i].display_info()
                found = True
            i += 1
        if not found:
            print(f"Pembeli {nama} tidak ditemukan.\n")

    # Mengubah data pembeli berdasarkan ID.
    def ubah_pembeli(self, id_pembeli, pembeli_baru):
        found = False
        i = 0
        while i < len(self.daftar_pembeli) and not found:
            if self.daftar_pembeli[i].get_id() == id_pembeli:
                self.daftar_pembeli[i] = pembeli_baru
                print("Pembeli berhasil diubah.\n")
                found = True
            i += 1
        if not found:
            print(f"Pembeli dengan ID {id_pembeli} tidak ditemukan.\n")

    # Menghapus pembeli berdasarkan ID.
    def hapus_pembeli(self, id_pembeli):
        found = False
        i = 0
        while i < len(self.daftar_pembeli) and not found:
            if self.daftar_pembeli[i].get_id() == id_pembeli:
                del self.daftar_pembeli[i]
                print("Pembeli berhasil dihapus.\n")
                found = True
            i += 1
        if not found:
            print(f"Pembeli dengan ID {id_pembeli} tidak ditemukan.\n")

    # ===== TRANSAKSI =====
    # Menambahkan transaksi baru, validasi pembeli dan produk, update stok.
    def tambah_transaksi(self, id_transaksi, id_pembeli, id_produk, tanggal, jumlah_item):
        # cari pembeli berdasarkan ID
        pembeli = None
        i = 0
        while i < len(self.daftar_pembeli) and pembeli is None:
            if self.daftar_pembeli[i].get_id() == id_pembeli:
                pembeli = self.daftar_pembeli[i]
            i += 1
        if not pembeli:
            print(f"Pembeli dengan ID {id_pembeli} tidak ditemukan.\n")
            return

        # cari produk berdasarkan ID
        produk = None
        i = 0
        while i < len(self.daftar_produk) and produk is None:
            if self.daftar_produk[i].get_id() == id_produk:
                produk = self.daftar_produk[i]
            i += 1
        if not produk:
            print(f"Produk dengan ID {id_produk} tidak ditemukan.\n")
            return

        # cek stok
        if produk.get_stok() < jumlah_item:
            print("Stok produk tidak mencukupi.\n")
            return

        # buat transaksi baru dan simpan
        total_harga = produk.get_harga() * jumlah_item
        transaksi = Transaksi(id_transaksi, id_pembeli, id_produk, tanggal, jumlah_item, total_harga)
        self.daftar_transaksi.append(transaksi)

        # update stok produk
        produk.set_stok(produk.get_stok() - jumlah_item)
        print("Transaksi berhasil ditambahkan.\n")

    # Menghapus transaksi berdasarkan ID.
    def hapus_transaksi(self, id_transaksi):
        found = False
        i = 0
        while i < len(self.daftar_transaksi) and not found:
            if self.daftar_transaksi[i].get_id_transaksi() == id_transaksi:
                self.daftar_transaksi.pop(i)
                print(f"Transaksi dengan ID {id_transaksi} berhasil dihapus.\n")
                found = True
            else:
                i += 1
        if not found:
            print(f"Transaksi dengan ID {id_transaksi} tidak ditemukan.\n")

    
    # Mengubah data transaksi berdasarkan ID.
    def update_transaksi(self, id_transaksi, transaksi_baru):
        found = False
        i = 0
        while i < len(self.daftar_transaksi) and not found:
            if self.daftar_transaksi[i].get_id_transaksi() == id_transaksi:
                self.daftar_transaksi[i] = transaksi_baru
                print(f"Transaksi dengan ID {id_transaksi} berhasil diubah.\n")
                found = True
            else:
                i += 1
        if not found:
            print(f"Transaksi dengan ID {id_transaksi} tidak ditemukan.\n")


    # Menampilkan semua transaksi (pakai for-each loop).
    def tampilkan_semua_transaksi(self):
        for t in self.daftar_transaksi:
            t.display_info()

File: java/Toko.java
import java.util.ArrayList;

public class Toko {
    // ===================== Atribut ===================== //
    private ArrayList<Produk> daftarProduk;
    private ArrayList<Pembeli> daftarPembeli;
    private ArrayList<Transaksi> daftarTransaksi;

    // ===================== Constructor ===================== //
    public Toko() {
        daftarProduk = new ArrayList<>();
        daftarPembeli = new ArrayList<>();
        daftarTransaksi = new ArrayList<>();
    }

    // ===================== BAGIAN PRODUK ===================== //
    // Menambahkan produk baru ke dalam daftar produk
    public void tambahProduk(Produk produk) {
        daftarProduk.add(produk);
        System.out.println("Produk berhasil ditambahkan.");
    }

    // Mencari produk berdasarkan nama
    public void cariProduk(String namaProduk) {
        boolean found = false;
        int i = 0;
        while (i < daftarProduk.size() && !found) {
            if (daftarProduk.get(i).getNama().equals(namaProduk)) {
                daftarProduk.get(i).displayInfo();
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Produk dengan nama " + namaProduk + " tidak ditemukan.");
        }
    }

    // Mengubah informasi produk berdasarkan ID
    public void ubahProduk(int id, Produk produkBaru) {
        boolean found = false;
        int i = 0;
        while (i < daftarProduk.size() && !found) {
            if (daftarProduk.get(i).getIdProduk() == id) {
                daftarProduk.set(i, produkBaru);
                System.out.println("Produk berhasil diubah.");
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Produk dengan ID " + id + " tidak ditemukan.");
        }
    }

    // Menghapus produk berdasarkan ID
    public void hapusProduk(int id) {
        boolean found = false;
        int i = 0;
        while (i < daftarProduk.size() && !found) {
            if (daftarProduk.get(i).getIdProduk() == id) {
                daftarProduk.remove(i);
                System.out.println("Produk berhasil dihapus.");
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Produk dengan ID " + id + " tidak ditemukan.");
        }
    }

    // Menampilkan semua produk yang ada di daftar produk
    public void tampilkanSemuaProduk() {
        // pakai for-each loop
        for (Produk produk : daftarProduk) {
            produk.displayInfo();
        }
    }

    // ===================== BAGIAN PEMBELI ===================== //
    // Menambahkan pembeli baru ke dalam daftar pembeli
    public void tambahPembeli(Pembeli pembeli) {
        daftarPembeli.add(pembeli);
        System.out.println("Pembeli berhasil ditambahkan.");
    }

    // Mencari pembeli berdasarkan nama
    public void cariPembeli(String namaPembeli) {
        boolean found = false;
        int i = 0;
        while (i < daftarPembeli.size() && !found) {
            if (daftarPembeli.get(i).getNamaPembeli().equals(namaPembeli)) {
                daftarPembeli.get(i).displayInfo();
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Pembeli dengan nama " + namaPembeli + " tidak ditemukan.");
        }
    }

    // Mengubah informasi pembeli berdasarkan ID
    public void ubahPembeli(int id, Pembeli pembeliBaru) {
        boolean found = false;
        int i = 0;
        while (i < daftarPembeli.size() && !found) {
            if (daftarPembeli.get(i).getIdPembeli() == id) {
                daftarPembeli.set(i, pembeliBaru);
                System.out.println("Data pembeli berhasil diubah.");
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Pembeli dengan ID " + id + " tidak ditemukan.");
        }
    }

    // Menghapus pembeli berdasarkan ID
    public void hapusPembeli(int idPembeli) {
        boolean found = false;
        int i = 0;
        while (i < daftarPembeli.size() && !found) {
            if (daftarPembeli.get(i).getIdPembeli() == idPembeli) {
                daftarPembeli.remove(i);
                System.out.println("Data pembeli berhasil dihapus.");
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Pembeli dengan ID " + idPembeli + " tidak ditemukan.");
        }
    }

    // Menampilkan semua pembeli yang ada di daftar pembeli
    public void tampilkanSemuaPembeli() {
        // pakai for-each loop
        for (Pembeli pembeli : daftarPembeli) {
            pembeli.displayInfo();
        }
    }

    // ===================== BAGIAN TRANSAKSI ===================== //
    // Menambahkan transaksi baru ke dalam daftar transaksi
    public void tambahTransaksi(int idTransaksi, int idPembeli, int idProduk, String tanggal, int jumlahItem) {
        boolean foundPembeli = false;
        int i = 0;
        // Cek apakah pembeli dengan idPembeli ada
        while (i < daftarPembeli.size() && !foundPembeli) {
            if (daftarPembeli.get(i).getIdPembeli() == idPembeli) {
                foundPembeli = true;
            }
            i++;
        }
        // Jika pembeli tidak ditemukan, keluar dari method
        if (!foundPembeli) {
            System.out.println("Pembeli dengan ID " + idPembeli + " tidak ditemukan.");
            return;
        }

        boolean foundProduk = false;
        int j = 0;
        // Cek apakah produk dengan idProduk ada
        while (j < daftarProduk.size() && !foundProduk) {
            // Jika produk ditemukan, cek apakah stok mencukupi
            if (daftarProduk.get(j).getIdProduk() == idProduk) {
                foundProduk = true;
                // Jika stok mencukupi, buat transaksi baru dan kurangi stok produk
                if (daftarProduk.get(j).getStok() >= jumlahItem) {
                    double totalHarga = daftarProduk.get(j).getHarga() * jumlahItem;
                    Transaksi transaksi = new Transaksi(idTransaksi, idPembeli, idProduk, tanggal, jumlahItem, totalHarga);
                    daftarTransaksi.add(transaksi);
                    daftarProduk.get(j).setStok(daftarProduk.get(j).getStok() - jumlahItem);
                } else {
                    System.out.println("Stok produk tidak mencukupi.");
                }
            }
            j++;
        }
        // Jika produk tidak ditemukan, tampilkan pesan
        if (!foundProduk) {
            System.out.println("Produk dengan ID " + idProduk + " tidak ditemukan.");
        }else {
            System.out.println("Transaksi berhasil ditambahkan.");
        }
    }

    // Menghapus transaksi berdasarkan ID transaksi
    public void hapusTransaksi(int idTransaksi) {
        boolean found = false;
        int i = 0;
        while (i < daftarTransaksi.size() && !found) {
            if (daftarTransaksi.get(i).getIdTransaksi() == idTransaksi) {
                daftarTransaksi.remove(i);
                System.out.println("Transaksi berhasil dihapus.");
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Transaksi dengan ID " + idTransaksi + " tidak ditemukan.");
        }
    }

    // Mengupdate informasi transaksi berdasarkan ID transaksi
    public void updateTransaksi(int idTransaksi, String tanggalBaru, int jumlahItemBaru) {
        boolean found = false;
        int i = 0;
        while (i < daftarTransaksi.size() && !found) {
            if (daftarTransaksi.get(i).getIdTransaksi() == idTransaksi) {
                Transaksi transaksi = daftarTransaksi.get(i);
                // Cari produk terkait untuk mengupdate stok
                boolean foundProduk = false;
                int j = 0;
                while (j < daftarProduk.size() && !foundProduk) {
                    if (daftarProduk.get(j).getIdProduk() == transaksi.getIdProduk()) {
                        foundProduk = true;
                        int selisihJumlah = jumlahItemBaru - transaksi.getJumlahItem();
                        // Cek apakah stok mencukupi untuk penambahan item
                        if (selisihJumlah > 0 && daftarProduk.get(j).getStok() < selisihJumlah) {
                            System.out.println("Stok produk tidak mencukupi untuk penambahan item.");
                            return;
                        }
                        // Update stok produk
                        daftarProduk.get(j).setStok(daftarProduk.get(j).getStok() - selisihJumlah);
                        // Update informasi transaksi
                        transaksi.setTanggal(tanggalBaru);
                        transaksi.setJumlahItem(jumlahItemBaru);
                        transaksi.setTotalHarga(daftarProduk.get(j).getHarga() * jumlahItemBaru);
                        System.out.println("Transaksi berhasil diupdate.");
                    }
                    j++;
                }
                found = true;
            }
            i++;
        }
        if (!found) {
            System.out.println("Transaksi dengan ID " + idTransaksi + " tidak ditemukan.");
        }
    }


    public void tampilkanSemuaTransaksi() {
        // pakai for-each loop
        for (Transaksi transaksi : daftarTransaksi) {
            transaksi.displayInfo();
        }
    }
}
